File: 프로그래머스/LV2/타겟넘버.py
from collections import deque
def solution(numbers, target):
    answer = 0
    def dfs(num, level):
        nonlocal answer
        if level == len(numbers):
            if num == target:
                answer += 1
            return
        dfs(num + numbers[level], level + 1)
        dfs(num - numbers[level], level + 1)
    def bfs():
        nonlocal answer
        q = deque()
        q.append([numbers[0], 1])
        q.append([-numbers[0], 1])
        while q:
            num, level = q.popleft()
            if level == len(numbers):
                if num == target:
                    answer += 1
            else:
                q.append([num + numbers[level], level + 1])
                q.append([num - numbers[level], level + 1])
    #dfs(numbers[0], 1)
    #dfs(-numbers[0], 1)
    bfs()
    return answer

File: 프로그래머스/LV2/타겟넘버.java
import java.util.*;
class Solution {
    public int solution(int[] numbers, int target) {
        int answer = bfs(numbers, target);
        //int answer = dfs(numbers, 0, 0, target);
        return answer;
    }
    
    private int dfs(int[] numbers, int num, int level, int target) {
        if(level==numbers.length) {
            return (num==target) ? 1 : 0;
        }
        return dfs(numbers, num+numbers[level], level+1, target) + dfs(numbers, num-numbers[level], level+1, target);
    }
    
    private int bfs(int[] numbers, int target) {
        int answer = 0;
        Queue<int[]> q = new LinkedList<>();
        q.add(new int[]{0, 0});
        while(q.size() > 0) {
            int[] element = q.poll();
            int num = element[0];
            int level = element[1];
            if(level == numbers.length) {
                if (num == target) answer++;
            }
            else {
                q.add(new int[]{num + numbers[level], level+1});
                q.add(new int[]{num - numbers[level], level+1});
            }
        }
        return answer;
    }
}
